### ISSUE TITLE
Automate the documentation of Autosubmit variables (for job, platform, and project) - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1828]
_Merges automate-variables-doc -> master_

Closes #951

Getting there. This draft produces the rst syntax that looks like the existing code.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/ec642d93f7ab1493b8e9a4e17a19a9d4/image.png)

(top is produced by the directive reading the code, bottom is the written docs text; section is duplicated as I hadn't time to complete that today).

The decorator added in this merge request populates a list of parameters. This list is used by the Sphinx directive to produce the documentation. This part is already unit-tested, 100% covered. The rest of the work before this is ready for review includes:

- [x] add a decorator to mark the parameters
- [x] use properties to define parameters
- [x] generate the three sections with variables for jobs, platforms, and projects, automatically
- [x] use the same list of parameters and docs to populate the comments in the YAML configuration where these parameters appear
- [x] add unit tests to the parts changed, or new code
- [x] make sure RTD rendered output matches the locally generated site - https://autosubmit--340.org.readthedocs.build/en/340/userguide/variables.html

**Comment by kinow** [comment]
p.s. this merge request includes more changes, mainly for testing. Not worth reviewing the code yet.

**Comment by kinow** [comment]
added 1 commit

<ul><li>56236298 - WIP adding some more variables and updating the markdown</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14612&start_sha=655da18bf9ffd122e1a846e9efc2d903c7dbf781)

**Comment by kinow** [comment]
added 1 commit

<ul><li>6798e66e - Parametrize the sphinx directive</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14614&start_sha=56236298ab4c17aef18a838e9687910ebc8df220)

**Comment by kinow** [comment]
added 1 commit

<ul><li>61a89ac1 - Fix typos, re-word text a bit</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14618&start_sha=6798e66eb5a2144d7ae801164aaa2d4e33eae868)

**Comment by kinow** [comment]
added 1 commit

<ul><li>2a90bf4d - Create decorators to meta-programmatically create the</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14626&start_sha=61a89ac1aa35f3ecf17ad3884a57b6d005ea89b5)

**Comment by kinow** [comment]
marked the checklist item **use properties to define parameters** as completed

**Comment by kinow** [comment]
marked the checklist item **generate the three sections with variables for jobs, platforms, and projects, automatically** as completed

**Comment by kinow** [comment]
Productive day! The automated documentation part is working for me. I reviewed the Sphinx formatting (things like VARIABLE are now displayed as `VARIABLE`), fixed some typos, added variables that were missing, and hooked Sphinx with the values created with the Python decorators.

Some values are computed on runtime, and cannot be moved to a `@property` at the moment. I think that could be refactored to have a `Job` object with less logic in its functions (i.e. instead of a long function that computes the parameters, isolate the logic of each parameter in `@property`). But that will be for another issue.

Next is to go through the templates in `autosubmit4-config-parser`. Find where these are used in Autosubmit (probably somewhere in the `expid` logic), and process the YAML file, augmenting it with the same text from the Variables page.

**Comment by kinow** [comment]
What is better, we can preview the documentation for this pull request :slight_smile: (after #878)

https://autosubmit--340.org.readthedocs.build/en/340/userguide/variables.html

It is missing some extra variables I created with a second decorator. I will investigate why later (it works on my local environment).

![Screen_Shot_2023-05-22_at_17.40.59-fullpage](https://earth.bsc.es/gitlab/es/autosubmit/uploads/3737bb63a66701ec756fbbc4f574633e/Screen_Shot_2023-05-22_at_17.40.59-fullpage.png)

**Comment by kinow** [comment]
Oh yeah, other changes in this merge request:

- Variables are now all in alphabetical order (it was hard to keep the previous order with the automation, although not impossible).
- All variables are caps-locks/upper case (before some had mixed case).
- The variable description always has the first letter in upper case, and ends with a period.
- Added some example syntax on how the variables can be used in the template scripts (preparing for #952).

**Comment by kinow** [comment]
>Next is to go through the templates in `autosubmit4-config-parser`. Find where these are used in Autosubmit (probably somewhere in the `expid` logic), and process the YAML file, augmenting it with the same text from the Variables page.

AS configuration is copied from the templates at `generate_as_config`: https://earth.bsc.es/gitlab/es/autosubmit/-/blob/9592607b231bb0e6e5a0f68ebf97d343dbc6f041/autosubmit/autosubmit.py#L1042

**Comment by kinow** [comment]
added 1 commit

<ul><li>5bcc3722 - Augment the copied/generated Autosubmit configuration with the comments for...</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14644&start_sha=2a90bf4d2abeb555c73f93aacb1febc5bdcb8c04)

**Comment by kinow** [comment]
`Tworks!

<details>
<summary>snippet used for testing - for posteriority</summary>

```python
# noinspection PyUnresolvedReferences
from autosubmit.job.job import Job
from autosubmit.helpers.parameters import PARAMETERS

from typing import Dict, List, Tuple, Union

from ruamel.yaml import YAML

import sys


def test1():
    origin = '/home/bdepaula/Development/python/workspace/autosubmit4-config-parser/autosubmitconfigparser/' \
             'config/files/local-minimal.yml'

    def add_comments_to_yaml(yaml_data, parameters, /, keys=None):
        """A recursive generator that visits every leaf node and yields the flatten parameter."""
        if keys is None:
            keys = []
        if isinstance(yaml_data, dict):
            for key, value in yaml_data.items():
                if isinstance(value, dict):
                    add_comments_to_yaml(value, parameters, [*keys, key])
                else:
                    parameter_key = '.'.join([*keys, key]).upper()
                    if parameter_key in parameters:
                        comment = parameters[parameter_key]
                        yaml_data.yaml_set_comment_before_after_key(key, before=comment, indent=yaml_data.lc.col)

    def recurse_into_parameters(parameters: Dict[str, Union[Dict, List, str]], keys=None) -> Tuple[str, str]:
        """Recurse into the ``PARAMETERS`` dictionary, and emits a dictionary.

        The key in the dictionary is the flattened parameter key/ID, and the value
        is the parameter documentation.

        :param parameters: Global parameters dictionary.
        :param keys: For recursion, the accumulated keys.
        :return: A dictionary with the
        """
        if keys is None:
            keys = []
        if isinstance(parameters, dict):
            for key, value in parameters.items():
                if isinstance(value, dict):
                    yield from recurse_into_parameters(value, [*keys, key])
                else:
                    key = key.upper()
                    # Here's the reason why ``recurse_into_yaml`` and ``recurse_into_parameters``
                    # are not one single ``recurse_into_dict`` function. The parameters have some
                    # keys that contain ``${PARENT}.key`` as that is how they are displayed in
                    # the Sphinx docs. So we need to detect it and handle it. p.s. We also know
                    # the max-length of the parameters dict is 2! See the ``autosubmit.helpers.parameters``
                    # module for more.
                    if not key.startswith(f'{keys[0]}.'):
                        yield '.'.join([*keys, key]).upper(), value
                    else:
                        yield key, value

    with open(origin) as f:
        parameter_comments = dict(recurse_into_parameters(PARAMETERS))

        # data = yaml.round_trip_load(f, preserve_quotes=True)
        yaml = YAML(typ='rt')
        data = yaml.load(f)
        add_comments_to_yaml(data, parameter_comments)
        yaml.dump(data, sys.stdout)

```
</details>

The current code creates an experiment with the following comments (most generated automatically from the code commetns):

**Exp**

```yaml
DEFAULT:
  # Job experiment ID.
  EXPID: "a008"
  # Default HPC platform name.
  HPCARCH: "local"
EXPERIMENT:
  # List of start dates
  DATELIST: '20000101'
  # List of members.
  MEMBERS: fc0
  # Unit of the chunk size. Can be hour, day, month, or year.
  CHUNKSIZEUNIT: month
  # Size of each chunk.
  CHUNKSIZE: '4'
  # Number of chunks of the experiment.
  NUMCHUNKS: '2'
  CHUNKINI: ''
  # Calendar used for the experiment. Can be standard or noleap.
  CALENDAR: standard
PROJECT:
  # Type of the project.
  PROJECT_TYPE: none
  # Folder to hold the project sources.
  PROJECT_DESTINATION: ''
GIT:
  PROJECT_ORIGIN: ''
  PROJECT_BRANCH: ''
  PROJECT_COMMIT: ''
  PROJECT_SUBMODULES: ''
  FETCH_SINGLE_BRANCH: true
SVN:
  PROJECT_URL: ''
  PROJECT_REVISION: ''
LOCAL:
  PROJECT_PATH: ''
PROJECT_FILES:
  FILE_PROJECT_CONF: ''
  FILE_JOBS_CONF: ''
  JOB_SCRIPTS_TYPE: ''
RERUN:
  RERUN: false
  RERUN_JOBLIST: ''
```

**autosubmit**

```yaml
CONFIG:
  # Current version of Autosubmit.
  AUTOSUBMIT_VERSION: "4.0.0b"
  # Maximum number of jobs permitted in the waiting status.
  MAXWAITINGJOBS: 20
  # Total number of jobs in the workflow.
  TOTALJOBS: 20
  SAFETYSLEEPTIME: 10
  RETRIALS: 0
#wrappers:
#  wrapper_sim:
#    TYPE: "vertical"
#    JOBS_IN_WRAPPER: "SIM"
MAIL:
  NOTIFICATIONS: False
  TO:
STORAGE:
  TYPE: pkl
  COPY_REMOTE_LOGS: true
```

These values match 100% what's in the Python code, and what is also in the generated Sphinx HTML docs (as this one also uses the code as source).

**Comment by kinow** [comment]
marked the checklist item **use the same list of parameters and docs to populate the comments in the YAML configuration where these parameters appear** as completed

**Comment by kinow** [comment]
added 1 commit

<ul><li>ab74e577 - Fix the build</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14652&start_sha=5bcc3722620fb41a304659e09be13130a839fa8d)

**Comment by kinow** [comment]
added 1 commit

<ul><li>2f56e032 - Fix the build</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14654&start_sha=ab74e5778133d8330838f478129d95af52fe32e2)

**Comment by kinow** [comment]
added 1 commit

<ul><li>a51a0473 - Fix the build</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14656&start_sha=2f56e032f6d980218f4c3b6e08858cc2408c7d50)

**Comment by kinow** [comment]
added 1 commit

<ul><li>128f88b5 - Fix the build</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14658&start_sha=a51a0473956c9440309e951f14b1518e97e5ad0d)

**Comment by kinow** [comment]
Fixed the build. Just pending to make sure the changes are well covered by tested and understand why RTD is not displaying the same output I have on my laptop, and then it will be ready to review.

**Comment by kinow** [comment]
added 1 commit

<ul><li>d2e5edd8 - Remove unused code and conf, add unit tests for the docs code</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14663&start_sha=128f88b5560382d43932edb0bfbfb05e6e390de4)

**Comment by kinow** [comment]
Removed some dead code in `job.py`, related to #1004

**Comment by kinow** [comment]
added 2 commits

<ul><li>fea3719f - Remove unused code and conf, add unit tests for the docs code</li><li>76527bec - Add unit test for the YAML comments</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14667&start_sha=d2e5edd89c8e44d6e88d00735d890dd5603cf1da)

**Comment by kinow** [comment]
marked the checklist item **add unit tests to the parts changed, or new code** as completed

**Comment by kinow** [comment]
Unit tests added, phew. Now looking at RTD.

**Comment by kinow** [comment]
There are no tests for the `dummy`, `minimal`, and this other part of the if/elif/else. I added one test for this last if branch, but in the future we need to add more tests (shouldn't be a problem for this MR as at least the test added is being tested now).

**Comment by kinow** [comment]
I used `.stem` here instead of `:-4` since some file in the future could have a different length extension (e.g. `yaml`, `docx`, etc.)

**Comment by kinow** [comment]
Removed some functions that were not used :thumbsup:

**Comment by kinow** [comment]
added 1 commit

<ul><li>4d435545 - Change RTD configuration to remove the ruamel.yaml CLIB package</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14673&start_sha=76527bec1a7bef12d5febc46abe8397f53571b36)

**Comment by kinow** [comment]
added 1 commit

<ul><li>1225f3a2 - Change RTD configuration to remove the ruamel.yaml CLIB package</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14676&start_sha=4d435545023d8dcd90e5d73d93ddff3a070d61c6)

**Comment by kinow** [comment]
marked the checklist item **make sure RTD rendered output matches the locally generated site - https://autosubmit--340.org.readthedocs.build/en/340/userguide/variables.html** as completed

**Comment by kinow** [comment]
setuptools is deprecated, default is pip

https://docs.readthedocs.io/en/stable/config-file/v2.html#packages

**Comment by kinow** [comment]
Reproduced the commands from RTD step by step, and got the same output, slightly different from what I had (it was missing some tables, produced with ruamel.yaml data).

Found a post on SO pointing out that was due to the older API of ruamel.yaml having changed. But since we were already using the new API in this MR (basically `from ruamel.yaml import YAML` and then `YAML.load`, instead of older `from ruamel.yaml import load, dump, etc.`), I couldn't find a reason for that not working.

I noticed ruamel.yaml brings the clib package. I believe that's to speed up the parsing of YAML files. I uninstalled it and then the issue was gone :confused: 

No explanation why, but I suspect it could be either due to the version of pip/setuptools on RTD, or some system dependency that is missing in RTD. For now the easiest fix I could find was to uninstall it and add a comment.

**Comment by kinow** [comment]
marked this merge request as **ready**

**Comment by kinow** [comment]
requested review from @dbeltrankyl

**Comment by kinow** [comment]
added 1 commit

<ul><li>a31f47b7 - Delete properties after they have been used (to save memory)</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14692&start_sha=1225f3a268461cbe215aa96ab70c14c71c9035c4)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 16:48

GReat work!

About the missing values and some clarifications: 

- DELAY: Affects the Job dependency, from which chunk the job will be considered in the workflow
- DELAY_RETRIALS: Time to wait for AS to resubmit a failed job
- EXPORT: Allows to send environment variables before submitting the job
- FREQUENCY: Affects job dependency. The job will be included in the workflow every $frequency chunks
- PACKED: Used internally to see if a job is inside a wrapper or not
- SYNCHRONIZE: <member/chunk/date> Affects to the current job, is equivalent to a "BARRIER" in OpenMP
- TASKTYPE: R,PYTHON, PYTHON2,BASH
- CURRENT_HYPERTHREADING: Enabled or disabled

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 16:52

Commented on [autosubmit/helpers/parameters.py line 8](https://github.com/BSC-ES/autosubmit/compare/93615fb5c4b956bfe732defe5d621cec95c4c03f..a31f47b78bf600dd91f769b705c51ee86ee1fabd#diff-df471bbbc8cbc9afb41e470b7e24c33bR8)



typo in "paramete"

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 16:59

Commented on [autosubmit/job/job.py line 174](https://github.com/BSC-ES/autosubmit/compare/93615fb5c4b956bfe732defe5d621cec95c4c03f..8c7b2e29ece2b04c1067107fdce73234e04be756#diff-f8751fbf5d2712919c09b25f56122c3fL174)



I used this function in the past for debugging which is why it is unused atm ( as the call is deleted ). 

Anyway, In 4.1 ( the memory optimization one ) I added` __str__` and `__repr__ `function that replaces this functionality

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 17:08

Commented on [autosubmit/job/job.py line 1333](https://github.com/BSC-ES/autosubmit/compare/93615fb5c4b956bfe732defe5d621cec95c4c03f..8c7b2e29ece2b04c1067107fdce73234e04be756#diff-f8751fbf5d2712919c09b25f56122c3fR1333)



Did you check that self.sdate is getting the correct value? Just in case

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 17:16

Commented on [.readthedocs.yaml line 20](https://github.com/BSC-ES/autosubmit/compare/93615fb5c4b956bfe732defe5d621cec95c4c03f..8c7b2e29ece2b04c1067107fdce73234e04be756#diff-0e9564ae09819e181ea0fc4a583c7bbbR20)



Nice find, I had an issue in lumi VM as well it doesn't work with the last ruamel version not sure why because it is working on my desktop. Anyway AS now is not using the last version but didn't think about the docs

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2023, 17:20

Thanks for the work and the effort! This was very necessary for the near future with so many new users, I think I revised all and did no find issues ( aside of minor comments)

**Comment by kinow** [comment]
Fixed, thanks!!!

**Comment by kinow** [comment]
changed this line in [version 16 of the diff](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14754&start_sha=a31f47b78bf600dd91f769b705c51ee86ee1fabd#fdf8c1cbc101f1d829e70bde47f2dc273dc96d31_8_8)

**Comment by kinow** [comment]
added 1 commit

<ul><li>8c7b2e29 - Add unit test for job sdate, fix typo (thanks @dbeltrankyl)</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/340/diffs?diff_id=14754&start_sha=a31f47b78bf600dd91f769b705c51ee86ee1fabd)

**Comment by kinow** [comment]
I had seen other tests were covering `sdate`, so I didn't test it. The `@property` field is calling the exact same thing as the code removed, so I assume it should work fine.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a78d009b213fa4af85d493ab931c3722/image.png)

But just to be on the safe side, I've added a test in `test_job.py` for that in the last commit:

```py
    def test_sdate(self):
        """Test that the property getter for ``sdate`` works as expected."""
        for test in [
            [None, None, ''],
            [datetime.datetime(1975, 5, 25, 22, 0, 0, 0, datetime.timezone.utc), 'H', '1975052522'],
            [datetime.datetime(1975, 5, 25, 22, 30, 0, 0, datetime.timezone.utc), 'M', '197505252230'],
            [datetime.datetime(1975, 5, 25, 22, 30, 0, 0, datetime.timezone.utc), 'S', '19750525223000'],
            [datetime.datetime(1975, 5, 25, 22, 30, 0, 0, datetime.timezone.utc), None, '19750525']
        ]:
            self.job.date = test[0]
            self.job.date_format = test[1]
            self.assertEquals(test[2], self.job.sdate)
```

**Comment by kinow** [comment]
Thanks for the review @dbeltrankyl. Feedback addressed, new test added, pipeline passed, and RTD preview is looking good :tada: 

I will create another MR later to fix the TODO's later with your suggestions.

Can you take a final look & approve if it's ready to be merged, please? Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 29, 2023, 10:04

> I will create another MR later to fix the TODO's later with your suggestions.

Fine for me 

> Can you take a final look & approve if it's ready to be merged, please? Thanks!

Looks good! thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 29, 2023, 10:04

approved this merge request

**Comment by kinow** [comment]
mentioned in commit 289366fdb8d222a83004cea33f6df456b50c1a44

**Comment by kinow** [comment]
Thank you!
